### ISSUE TITLE
Ouroboros is unsound

### ISSUE THREAD

**Issue by GRA0007** [SeaQL/sea-orm#1705]
The creator of Ouroboros has just posted about the unsoundness of ouroboros, and indicated that if you're currently using it you should migrate to [self_cell](https://github.com/Voultapher/self_cell) instead where possible.

GitHub issue: https://github.com/joshua-maros/ouroboros/issues/88
Security advisory: https://github.com/advisories/GHSA-87mf-9wg6-ppf8

**Comment by billy1624** [comment]
Hey @GRA0007, thanks for letting us know. We will investigate it right now. And then perform the migration :)

**Comment by tyt2y3** [comment]
To be precise, ouroboros is only used in our `stream()` interface.

In short, we need a way to tell the compiler that the Stream depends on the Connection and it cannot outlive the connection.

I reckon, an ugly solution might be to use something like https://docs.rs/async-stream/latest/async_stream/ to implement the Stream interface? But then we'd have to Box them once again.

@nappa85 what's your thought?

**Comment by nappa85** [comment]
`ouroboros` is neede because the stream contains also the connection it originates from, hence it's a self-referential struct.

I don't know `async_stream`, but reading the docs it seems to use a channel to send stream results to the awaiter.

Just writing my thoughts...

It would be easy to do the same with `tokio::spawn` and a channel that implements Stream, but I think we want to avoid the channel to have a buffer, or we could end up having the entire Stream contents allocated inside the buffer. We should find a channel type that, on send, awaits that the single slot buffer is emptied by the receiver

Could spawning the stream in a separate task cause problems in some situations? Do we still need to be compatible with `async-std`?

**Comment by billy1624** [comment]
I just tried `self_cell`. I was blocked by the fact that I cannot take ownership of the connection pool. It used to be possible in `ouroboros`

```rs
impl QueryStream {
    #[instrument(level = "trace", skip(metric_callback))]
    fn build(
        stmt: Statement,
        conn: InnerConnection,
        metric_callback: Option<crate::metric::Callback>,
    ) -> QueryStream {
        QueryStream::new(
            QueryStreamInner(stmt, conn, metric_callback),
            |QueryStreamInner(stmt, conn, _metric_callback)| match conn { // This `conn` is borrowed (in "self_cell") not owned (in "ouroboros")
                #[cfg(feature = "sqlx-mysql")]
                InnerConnection::MySql(c) => {
                    let query = crate::driver::sqlx_mysql::sqlx_query(stmt);
                    let _start = _metric_callback.is_some().then(std::time::SystemTime::now);
                    let stream = c // We need to take ownership here
                        .fetch(query)
                        .map_ok(Into::into)
                        .map_err(sqlx_error_to_query_err);
                    let elapsed = _start.map(|s| s.elapsed().unwrap_or_default());
                    MetricStream::new(_metric_callback, stmt, elapsed, stream)
                }
```

**Comment by nappa85** [comment]
Ok, with runtime abstraction and flume we should be able to do something like this (really high level):
```rust
        let (tx, rx) = flume::bounded(0);
        spawn(async move {
                let query = crate::driver::sqlx_mysql::sqlx_query(stmt);
                let stream = conn
                    .fetch(query)
                    .map_ok(Into::into)
                    .map_err(sqlx_error_to_query_err);
                while let Some(row) = stream.next().await {
                  if rx.send_async(row).is_err() {
                    break;
                  }
                }
            }
          }
        });
        rx.into_stream()
```

Main problem I see from that snippet is stat we lose metric support, I need to better think about it...

I'm open to proposals for a better solution

**Comment by billy1624** [comment]
Seems `ouroboros` team found a solution to the unsound issue? https://github.com/joshua-maros/ouroboros/issues/88#issuecomment-1592123323

**Comment by tyt2y3** [comment]
Well, that seems to be a good news! Let me re-read the issue and see if this issue may be closeable too (at least we have to migrate to 0.16)

**Comment by nappa85** [comment]
It would be useful to know if the unsoundness has hit us, because I never had a single problem with our streams, so or the bug is really difficult to trigger, or we were using the crate the right way and we weren't affected

**Comment by tyt2y3** [comment]
Ref: https://github.com/SeaQL/sea-orm/security/dependabot/2
